### ISSUE TITLE
Following error is coming while importing cvxopt even after successful installation  using pip

### ISSUE THREAD

**Issue by ashwinirahangdale** [cvxopt/cvxopt#127]
I have installed cvxopt using the pip. It has been installed successfully but while importing this package we are getting following error.

 import cvxopt
	     
Traceback (most recent call last):
  File "<pyshell#28>", line 1, in <module>
    import cvxopt
  File "C:\Users\Ashwini\AppData\Local\Programs\Python\Python36\lib\site-packages\cvxopt\__init__.py", line 50, in <module>
    import cvxopt.base
ImportError: DLL load failed: The specified module could not be found.

**Comment by ashwinirahangdale** [comment]
I am using Python3.6+.
 I tried to include that contain the MKL in my path. It may be possible that I am not adding it correctly. Could you please explain how to include the path of MKL directory properly?

**Comment by ashwinirahangdale** [comment]
I am using Python3.6+.
 I tried to include that contain the MKL in my path. It may be possible
that I am not adding it correctly. Could you please explain how to include
the path of MKL directory properly?
ASHWINI RAHANGDALE
Ph.D. Research Scholar
VNIT Nagpur



[image: Mailtrack]
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
Sender
notified by
Mailtrack
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
09/08/18,
16:38:50

**Comment by martinandersen** [comment]
Open the *System Properties* window and click on the *Advanced* tab. Click the *Environment Variables* button at the bottom. In the *User variables* section, select *Path* and click *Edit*. Add the directory that contains `mkl_rt.dll` to the path. You will probably need to restart Python after updating the path.

Are you using Python from Python.org or some other Python distribution?

**Comment by ashwinirahangdale** [comment]
It is working.  ow, not getting any error for importing cvxopt. Thank you
so much for your valuable response.
ASHWINI RAHANGDALE
Ph.D. Research Scholar
VNIT Nagpur


[image: Mailtrack]
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
Sender
notified by
Mailtrack
<https://mailtrack.io?utm_source=gmail&utm_medium=signature&utm_campaign=signaturevirality5&>
10/08/18,
11:07:27

**Comment by martinandersen** [comment]
I'm trying to reproduce the problem. How did you install Python? Did you download the installer from python.org or did you install some other Python distribution?

**Comment by martinandersen** [comment]
I'm closing this issue since I cannot reproduce the problem.

**Comment by wkschwartz** [comment]
I also ran into this problem. Prepending my `PATH` with the directory containing `mkl_rt.dll` worked, as @martinandersen suggested above (https://github.com/cvxopt/cvxopt/issues/127#issuecomment-411973792).

Maybe the reason @martinandersen couldn't reproduce is MKL is already available on his `PATH`? MKL was not already available on my path and I only had it installed at all in my [venv](https://docs.python.org/3/library/venv.html) where I had installed cvxopt.

It would be a nasty surprise for library code such as cvxopt to manipulate the PATH. Perhaps it can call the DLL with an absolute path, like `os.path.join(sys.prefix, 'Library', 'bin') if os.name == 'nt' else ...`?

**Comment by wkschwartz** [comment]
On my Mac, it appears there's not equivalent directory containing MKL `.so`s, so maybe the DLL-loading code is pretty Windows specific?

**Comment by martinandersen** [comment]
It is definitely Windows specific. 

The macOS wheels use Accelerate instead of MKL since the MKL wheel for macOS will not work properly unless you manually make some changes (the Anaconda version of MKL includes this fix: [ref](https://github.com/AnacondaRecipes/aggregate/commit/91c5d09677e3d247b31572fb29081f2bc91d96c9)). This has nothing to do with the path.

Maybe there's a difference between Anaconda Python and the official Python in terms of search path on Windows?

**Comment by wkschwartz** [comment]
Where is the code that loads `mlk_rt`? (I can't find code in this repo that actually refers to `mkl_rt`, or anything else involving `dlopen` that looked related. I'm sure I'm just using the wrong search terms.)

If you can help me figure out where to patch, I'll upstream whatever I come up with. I need to get CVXOPT working for a current project, and munging the PATH will probably break other things in my app. (We're a Pip shop, so using Anaconda isn't really an option.)

**Comment by martinandersen** [comment]
There is no such code in CVXOPT. We rely on dynamic linking on all platforms.

It is a bit strange to me that Python doesn't include `%PREFIX%\Library\bin` in its search path. I suppose Anaconda does this.

Perhaps you could use `sys.path.append()` as a workaround?

**Comment by wkschwartz** [comment]
I'll try `sys.path` as a workaround for now (will report back here if it works). It's still a workaround that I would need to `try ... finally` guard so it doesn't mess with other code (e.g., causing `import` to grab the wrong module somewhere else in my code). But at least it's easier to reason about than the system PATH.

I need to correct myself: I think that it should actually be `sys.exec_prefix`, not `sys.prefix`. The [docs](https://docs.python.org/3/library/sys.html#sys.exec_prefix) say that `exec_prefix` is where platform _dependent_ code goes.

After poking around a bit, I think the underlying problem is that `Library\bin` is a non-standard directory name in the Python world. The link above and [this one to Python's docs](https://docs.python.org/3/using/windows.html#finding-modules) seem to indicate that library code is stored in `Lib` on Windows:
>Python usually stores its library (and thereby your site-packages folder) in the installation directory. So, if you had installed Python to `C:\Python\`, the default library would reside in `C:\Python\Lib\` and third-party modules should be stored in `C:\Python\Lib\site-packages\`.

Here, `C:\Python` is the prefix, so that changes if you're in a venv.

Looking at the installation of OSQP and CVXPY, they store their dynamic libraries as `.pyd` files in `Lib\site-packages` and `Lib\site-packages\<package name>`.

Is it possible to change where `mkl_rt.dll` is stored? Since I can't find `"mkl_rt"` anywhere in the source code, I can't figure out where the wheel builder/installer decides where to put the file. Any hints? I know it's all done through dynamic linking, but there must be somewhere that says what to link to, right?

**Comment by wkschwartz** [comment]
Calling the solver in my application happens in a `solve` method in some class. Neither of the following worked while overriding `solve`:

```python
def solve(self):
    if os.name == 'nt':
        lib_bin = os.path.join(sys.exec_prefix, 'Library', 'bin')
        old_path = os.environ['PATH']
        os.environ['PATH'] = os.pathsep.join([lib_bin, old_path])
        try:
            return super().solve()
        finally:
            os.environ['PATH'] = old_path
    else:
        return super().solve()
```

```python
def solve(self):
    if os.name == 'nt':
        lib_bin = os.path.join(sys.exec_prefix, 'Library', 'bin')
        old_path = sys.path
        sys.path.insert(0, lib_bin)
        try:
            return super().solve()
        finally:
            sys.path = old_path
    else:
        return super().solve()
```
The only thing that works so far is setting the path by hand (either through the control panel or using the `set PATH=` command in `cmd.exe`) before running the code.

**Comment by martinandersen** [comment]
You can see how we build the Windows wheels here:
https://github.com/cvxopt/cvxopt-wheels/blob/master/.appveyor.yml

Basically, we add MKL as a dependency via the environment variable `CVXOPT_INSTALL_REQUIRES=mkl` so that pip will install MKL before installing CVXOPT. The MKL wheel (https://pypi.org/project/mkl/) determines the location of `mkl_rt`. 

The primary reason for using MKL on Windows is that it is (as far as I know) the only BLAS/LAPACK library that is available as a wheel and may be installed automatically as a dependency. However, if this is not a reliable solution, then we may need to reconsider.

**Comment by wkschwartz** [comment]
Thanks. That information led me to a solution that involves temporarily patching PATH, using ctypes.windll.LoadLibrary to load `mkl_rt` and some others it depends on, and then un-doing the change to PATH. This has to happen before importing cvxopt. Currently I've applied this patch to my code. I'll send a PR to apply this patch to cvxopt when I get some time in a couple of days.

Upshot: MKL can continue to work, but some allowance needs to be made for standard (i.e., non-conda) installations of Python.

**Comment by wkschwartz** [comment]
For my reference later (and so I can close these tabs in my browser), here are some links I used in my research:
- https://conda.io/docs/user-guide/tasks/build-packages/use-shared-libraries.html
- https://github.com/ContinuumIO/anaconda-issues/issues/509
- https://github.com/cvxopt/cvxopt/blob/master/setup.py
- https://mail.python.org/pipermail/distutils-sig/2014-October/024993.html
- https://mail.python.org/pipermail/distutils-sig/2014-September/024983.html
- The thread containing the above two links is useful as a whole.

**Comment by martinandersen** [comment]
Thanks for looking into this, @wkschwartz. 

The most recent MKL wheel package (2019.0) for Windows leaves out some .lib files that appear to be necessary at link time when building CVXOPT, so we have had to pin the MKL version in the [cvxopt-wheels](https://github.com/cvxopt/cvxopt-wheels) repository. This is probably the appropriate place to apply a Windows-specific patch. 

The numpy wheels for Windows appear to include OpenBLAS, so an alternative would be to make Numpy a dependency on Windows instead of MKL.

**Comment by wkschwartz** [comment]
Sorry I dropped the ball on this. Keeping NumPy as a dependency sounds way easier than MKL, but I don't know enough about MKL's vs. OpenBLAS's support and performance to comment on any trade offs between them.

Here's a bootstrapping function (slightly modified for public consumption) that I created for my project that depends on CVXOPT.


```python
# _mkl_bootstrap.py
import os
import threading


_mkl_loaded = threading.Event()


def _load_mkl_win() -> None:  # pragma: no cover
	r"""Load, if necessary, Intel's Math Kernel Library (MKL).

	First, why might we want MKL, and second, when is it necessary to load it
	now?

	CVXOPT requires MKL to implement BLAS, which is a set of industry-standard
	linear algebra subroutines. CVXOPT only requires MKL on Windows because
	OpenBLAS is generally available on Linux and macOS comes with its own
	Accelerate Framework implementation of BLAS. Intel provides MKL as a Wheel
	on PyPI, which Pip installs as part of installing CVXOPT. See `#127
	<https://github.com/cvxopt/cvxopt/issues/127>`_.

	Intel's MKL wheel places a bunch of DLLs in the
	:file:`{exec_prefix}\Library\bin` directory, where :file:`{exec_prefix}` is
	:const:`sys.exec_prefix`. It chooses this location because the Conda package
	manger, which is popular in the scientific community but is not standard in
	the larger Python community, `uses this directory on Windows for DLLs
	<https://conda.io/docs/user-guide/tasks/build-packages/use-shared-libraries.html#shared-libraries-in-windows>`_.
	If you run Conda's version of Python, then the MKL wheel works out of the
	box. If you run standard Python, you're in trouble.

	This is because Windows loads DLLs from the system PATH environment
	variable. :file:`{exec_prefix}\Library\bin` is not on the PATH by default.
	We could add it to the PATH when this module loads, but that's a side effect
	of loading a library module that client code should not have to plan around
	(e.g., client code might load this module and then try to run a system
	command and be surprised to end up running a different command only because
	it's in :file:`{exec_prefix}\Library\bin`).

	Fortunately, Windows only loads DLLs once. So all we have to do is load MKL
	*before* importing CVXPY (which in turn imports CVXOPT) with
	:file:`{exec_prefix}\Library\bin` on the PATH, and then remove it from the
	PATH. That's what this function does.

	:raises ImportError:
		If MKL cannot be found.
	"""
	# Only needed on Windows.
	if os.name != 'nt':
		return
	# Only needed once per process
	if  _mkl_loaded.is_set():
		return
	# cdll vs windll doesn't matter because we're not actually calling any
	# functions. We use cdll here because Mypy complains about windll's
	# unavailability on non-Windows systems.
	from ctypes import cdll
	import sys
	# Use sys.exec_prefix because that's the home of platform-dependent files.
	lib_bin = os.path.join(sys.exec_prefix, 'Library', 'bin')
	# These are the two DLLs that Windows complains about if not preloaded.
	libs = ['mkl_rt', 'mkl_intel_thread']
	libs = [os.path.join(lib_bin, name + os.extsep + 'dll') for name in libs]
	old_path = os.environ['PATH']
	os.environ['PATH'] = os.pathsep.join([lib_bin, old_path])
	try:
		for lib in libs:
			try:
				cdll.LoadLibrary(lib)
			except OSError as e:
				raise ImportError(
					"Cannot find Intel's Math Kernel Library (mkl)") from e
	else:
		_mkl_loaded.set()
	finally:
		os.environ['PATH'] = old_path
```

To use it properly, before importing anything that needs MKL, put

```python
from ._mkl_bootstrap import _load_mkl_win
_load_mkl_win()
del _load_mkl_win
```

You only need to do that at most once per process. I'm not sure if child processes inherit the parent's DLL cache, obviating the need to re-bootstrap MKL in the child process.

If you use flake8, you may want to toss a [`# noqa: E402`](https://pycodestyle.readthedocs.io/en/latest/intro.html#error-codes) in there somewhere.

**Comment by wkschwartz** [comment]
Another option would be to vendor MKL in the Windows wheel. Then you'd need a boilerplate Python DLL to link to MKL, because Python ensures that 3rd-party DLLs in an importable directory next to Python DLLs are dynamically linkable at run time. I think I saw that [here](https://docs.python.org/3/extending/windows.html#differences-between-unix-and-windows) but quick skimming isn't helping me re-find it.

**Comment by martinandersen** [comment]
Thanks for sharing your bootstrapping function @wkschwartz.

**Comment by friederikemeier** [comment]
Another easy way is to use the packages for cvxopt and numpy+mkl from this site:
https://www.lfd.uci.edu/~gohlke/pythonlibs/#numpy
https://www.lfd.uci.edu/~gohlke/pythonlibs/#cvxopt

**Comment by wkschwartz** [comment]
@martinandersen I'm not sure if you ever ended up doing anything with the DLL loading code, but just wanted to bring to your attention that Python 3.8 will have [an official solution now](https://docs.python.org/3.8/whatsnew/3.8.html#bpo-36085-whatsnew)

**Comment by martinandersen** [comment]
Thanks @wkschwartz!

**Comment by neelotpal1112** [comment]
Is there a resolution for this apart from using Gohkle's website as suggested by @friederikemeier? 

I have a conda distribution of python (3.6) and wanted to install both cvxopt and statsmodels in the same virtual environment. Having a strict priority to use only conda-forge results in unresolvable conflicts. 

So I let conda install statsmodels and all the packages it depends on (including mkl) and installed cvxopt from the wheel on Golhke's website. I then prepended the mkl package folder to my PATH. I can verify that both mkl_rt.dll and mkl_intel_thread.dll are in this folder. Despite all of this, I am still running into this problem when I try to import cvxopt.

Note that the mkl version is 2019.4 so maybe there are other dll files missing?

Attached is the list of packages in this virtual environment in case it helps.

[python packages.txt](https://github.com/cvxopt/cvxopt/files/3735793/python.packages.txt)

**Comment by cgohlke** [comment]
> Note that the mkl version is 2019.4 so maybe there are other dll files missing?

The cvxopt builds at https://www.lfd.uci.edu/~gohlke/pythonlibs/#cvxopt do not use or depend on Intel MKL. That would violate the GPL 3 license AFAICT. Instead those builds depend on open source `libopenblas.dll`, which is also included in the numpy+MKL wheels and can be extracted manually from the wheels (which are ZIP files) if necessary...

**Comment by martinandersen** [comment]
I believe that `pip install cvxopt` should work without messing with paths if you are using anaconda Python. AFAIK, this issue only pertains to the official Python distribution for Windows.

**Comment by neelotpal1112** [comment]
Thanks, @martinandersen. A simple pip install cvxopt worked. Thanks, @cgohlke for that clarification.

**Comment by mborysiak** [comment]
> 
> 
> I believe that `pip install cvxopt` should work without messing with paths if you are using anaconda Python. AFAIK, this issue only pertains to the official Python distribution for Windows.

Will using the pip install directly on anaconda distribution with Windows allow for the glpk solver to be used?

**Comment by martinandersen** [comment]
@mborysiak, yes, the Windows wheel does include the GLPK extension.

**Comment by martinandersen** [comment]
We have added Windows wheels (x86 and x64) for Python 3.5-3.8 with OpenBLAS. Wheels for Python 3.9 will be available once AppVeyor adds support for Python 3.9.
